packages/parameters/src/BaseProvider.ts

```diff
@@ -0,0 +1,214 @@
+  public forceFetch: boolean = false;
+  public maxAge: number = DEFAULT_MAX_AGE_SECS;
+  public sdkOptions?: unknown;
+  public throwOnTransformError: boolean = false;
+  public transform?: TransformOptions;
+
+  public constructor(options: GetMultipleOptionsInterface) {
+    Object.assign(this, options);
+  }
+}
+
+class ExpirableValue implements ExpirableValueInterface {
+  public ttl: number;
+  public value: string | Record<string, unknown>;
+
+  public constructor(value: string | Record<string, unknown>, maxAge: number) {
+    this.value = value;
+    const timeNow = new Date();
+    this.ttl = timeNow.setSeconds(timeNow.getSeconds() + maxAge);
+  }
+
+  public isExpired(): boolean {
+    return this.ttl < Date.now();
+  }
+}
+
+abstract class BaseProvider implements BaseProviderInterface {
+  public store: Map<string, ExpirableValue> = new Map;
+
+  public constructor () {
+    this.store = new Map();
+  }
+
+  public addToCache(key: string, value: string | Record<string, unknown>, maxAge: number): void {
+    if (maxAge <= 0) return;
+
+    this.store.set(key, new ExpirableValue(value, maxAge));
+  }
+
+  public clearCache(): void {
+    this.store.clear();
+  }
+  
+  /**
+   * Retrieve a parameter value or return the cached value
+   * 
+   * If there are multiple calls to the same parameter but in a different transform, they will be stored multiple times.
+   * This allows us to optimize by transforming the data only once per retrieval, thus there is no need to transform cached values multiple times.
+   * 
+   * However, this means that we need to make multiple calls to the underlying parameter store if we need to return it in different transforms.
+   * 
+   * Since the number of supported transform is small and the probability that a given parameter will always be used in a specific transform,
+   * this should be an acceptable tradeoff.
+   * 
+   * @param {string} name - Parameter name
+   * @param {GetOptionsInterface} options - Options to configure maximum age, trasformation, AWS SDK options, or force fetch
+   */
+  public async get(name: string, options?: GetOptionsInterface): Promise<undefined | string | Record<string, unknown>> {
+    const configs = new GetOptions(options);
+    const key = [ name, configs.transform ].toString();
+
+    if (!configs.forceFetch && !this.hasKeyExpiredInCache(key)) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      return this.store.get(key)!.value;
+    }
+
+    let value;
+    try {
+      value = await this._get(name, options?.sdkOptions);
+    } catch (error) {
+      throw new GetParameterError((error as Error).message);
+    }
+
+    if (value && configs.transform) {
+      value = transformValue(value, configs.transform, true);
+    }
+
+    if (value) {
+      this.addToCache(key, value, configs.maxAge);
+    }
+
+    // TODO: revisit return type once providers are implemented, it might be missing binary when not transformed
+    return value;
```

Comment:
As mentioned in the comment here, I stayed generic with types because at this stage I'm not yet sure what each provider/AWS SDK (from `_get`) will return.

I have opened #1172 to track this work.